### PR TITLE
Add TypeScript definitions for v15.0.0.

### PR DESCRIPTION
**Pull request by parkerziegler**
**This pull request adds TypeScript definitions for v15.0.0 of `redux-little-router`.** It focuses on typing the high level aspects of the API, including `actions`, the `Location` and `History` interfaces, `routers`, and the `Link` and `Fragment` components. There is a demo repository created [here](https://github.com/parkerziegler/rlr-ts) to demonstrate using `redux-little-router` with TypeScript. The demo repository copies over the `demo/client` app, successfully passing the TS compiler. To run the demo locally, follow the instructions in the [README](https://github.com/parkerziegler/rlr-ts/blob/master/README.md).

screen shot 2018-02-11 at 9 49 42 am
screen shot 2018-02-11 at 9 50 11 am

Type definitions were also created for server-side rendering, but have not been tested against a demo app. This will be important to do before shipping a new release.

Unfortunately, the types submitted to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/redux-little-router/index.d.ts) don't fully support the API in its current state. Running the demo repository against these types results in the following Terminal output:
screen shot 2018-02-11 at 10 12 43 am

I am concerned that folks using the API with the definitions on DefintelyTyped may run into issues where parts of our API that are supported in JS will not pass the TS compiler because of incorrect typings. Shipping our own TypeScript definitions with `redux-little-router` ensures we keep them current and consistent with the latest version of our API. I would love to have some folks review these types, critique them, and expand them if possible.

### CODE REVIEW

**Comment by parkerziegler** [comment]
@tptee @ryan-roemer I think I'll need to add in defs for the `immutable` aspects of the API so the compiler won't yell about `import`s, I'll make the fix and push to this PR.

**Comment by parkerziegler** [comment]
@tptee @ryan-roemer Types have been added for the immutable aspects of the API. There is also a corresponding branch in the [`rlr-ts` repo](https://github.com/parkerziegler/rlr-ts), [`immutable-demo`](https://github.com/parkerziegler/rlr-ts/tree/immutable-demo), that verifies the types pass the compiler. `webpack` is correctly resolving the pointer to the ES build as well (`redux-little-router/es/immutable/index.js`). Let me know how I can help you out further!

screen shot 2018-02-12 at 9 53 24 pm

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  type: string;
+  payload: Location;
+};
+
+export function initializeCurrentLocation(location: Location): {
+  type: string;
+  payload: Location;
+};
+
+export function replaceRoutes(routes: {}): {
```
Comment (tptee) [inline index.d.ts]: TS noob question, can the `routes` conform to the `Routes` interface above?

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+export interface Router {
+  reducer: Reducer<Location>;
+  middleware: Middleware;
+  enhancer: StoreEnhancer<Location>;
+}
+
+export interface BrowserRouterArgs {
+  routes: {};
```
Comment (tptee) [inline index.d.ts]: Here as well

File: index.d.ts (hunk 1 above)
Comment (parkerziegler) [reply index.d.ts]: Good catch, we should be able to do that!

File: index.d.ts (hunk 2 above)
Comment (parkerziegler) [reply index.d.ts]: Yep!

File: index.d.ts (hunk 1 above)
Comment (parkerziegler) [reply index.d.ts]: @tptee I'll get on this later tonight and plan to merge tomorrow. Need to do some introspection around [index signatures](https://basarat.gitbooks.io/typescript/docs/types/index-signatures.html). I'll ping you when things are ready!

**Comment by parkerziegler** [comment]
@tptee We should be good to merge! I tightened up those definitions – they actually helped me to uncover an issue from the old typings that was over-restrictive on the data structures we could store in the `routes` object. Good eye! 🎉 💯

**Pull request merged**